### PR TITLE
Java Solution for Good Subarrays

### PR DESCRIPTION
**Pull request by brad182**
added comments to the c++ code as well

### CODE REVIEW

File: solutions/cf-1398C.mdx [hunk 1]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
 	map<int, ll> sumDist;
-	for(int i=0; i<=n; i++){
+	for(int i=0; i<=n; i++){  // build map
 		sumDist[dp[i]-i]++;
 	}
 	ll goodArrays=0;
 	for(pi p: sumDist){
 		ll f=p.second;
+		// this uses the formula n * (n - 1) / 2, and it gets
+            	// the sum of all the positive integers up to and including n
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:39]: not including n

File: solutions/cf-1398C.mdx [hunk 2]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
 void solve(){
 	int n; cin >> n;
 	vector<int> dp(n+1, 0);
-	for(int i=1; i<=n; i++){
+	for(int i=1; i<=n; i++){  // put into array
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:24]: this comment seems unnecessary, it's not necessary to restate what the code is doing

File: solutions/cf-1398C.mdx [hunk 3]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
 		char c; cin >> c;
 		dp[i]=c-'0';
 	}
-	for(int i=1; i<=n; i++){
+	for(int i=1; i<=n; i++){  // build prefix sums
 		dp[i]+=dp[i-1];
 	}
 	map<int, ll> sumDist;
-	for(int i=0; i<=n; i++){
+	for(int i=0; i<=n; i++){  // build map
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:32]: same as first comment

File: solutions/cf-1398C.mdx [hunk 4]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
+            	// the sum of all the positive integers up to and including n
 		goodArrays+=f*(f-1)/2;
 	}
 	cout << goodArrays << endl;
 }
 
 int main(){
 	int t; cin >> t;
-	for(int i=0; i<t; i++){
+	for(int i=0; i<t; i++){  // each test case
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:47]: same as first comment

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/cf-1398C.mdx [hunk 5]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
 void solve(){
 	int n; cin >> n;
 	vector<int> dp(n+1, 0);
-	for(int i=1; i<=n; i++){
+	for(int i=1; i<=n; i++){
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:24]: remove these extra spaces

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/cf-1398C.mdx [hunk 6]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
+			}
+			else {  // not in the map yet, put 1
+				sumDist.put(array[x] - x, (long) 1);
+			}
+		}
+		long goodArrays = 0;
+		for (int key : sumDist.keySet()) {  // for each key
+			long f = sumDist.get(key);
+			// this uses the formula n * (n - 1) / 2, and it gets
+			// the sum of all the positive integers up to n
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:84]: "1 up to n" is ambiguous - it could refer to 1...n

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/cf-1398C.mdx (hunk 6 above)
Comment (jessechoe10) [reply solutions/cf-1398C.mdx:84]:
```suggestion
			// the sum of the first n positive integers
```

**Review by jessechoe10 — CHANGES_REQUESTED** [review]

File: solutions/cf-1398C.mdx (hunk 6 above)
Comment (bqi343) [reply solutions/cf-1398C.mdx:84]: it's the first n-1 positive integers though

File: solutions/cf-1398C.mdx (hunk 5 above)
Comment (brad182) [reply solutions/cf-1398C.mdx]: i can't really see them, don't know if I've missed any

File: solutions/cf-1398C.mdx [hunk 7]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
 		char c; cin >> c;
 		dp[i]=c-'0';
 	}
-	for(int i=1; i<=n; i++){
+	for (int i=1; i<=n; i++) {
 		dp[i]+=dp[i-1];
 	}
 	map<int, ll> sumDist;
-	for(int i=0; i<=n; i++){
+	for (int i=0; i<=n; i++) {
```
Comment (maggie-j-liu) [inline solutions/cf-1398C.mdx:32]:
```suggestion
	for (int i=0; i<=n; i++) {
```

File: solutions/cf-1398C.mdx [hunk 8]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
 		goodArrays+=f*(f-1)/2;
 	}
 	cout << goodArrays << endl;
 }
 
-int main(){
+int main() {
 	int t; cin >> t;
-	for(int i=0; i<t; i++){
+	for(int i=0; i<t; i++){
```
Comment (maggie-j-liu) [inline solutions/cf-1398C.mdx:47]:
```suggestion
	for (int i=0; i<t; i++) {
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Review by jessechoe10 — APPROVED** [review]

File: solutions/cf-1398C.mdx [hunk 9]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
+		for (int x = 1; x <= arrayLength; x++) {  
+			array[x] = stringArray.charAt(x - 1) - '0';
+		}
+		for (int x = 1; x <= arrayLength; x++) { 
+			array[x] += array[x - 1];
+		}
+		Map<Integer, Long> sumDist = new HashMap<>();
+		for (int x = 0; x <= arrayLength; x++) {  
+			if (sumDist.containsKey(array[x] - x)) {  // already in map
+				// add one on to what is already there
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:73]: redundant comments

File: solutions/cf-1398C.mdx [hunk 10]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
+				sumDist.put(array[x] - x, sumDist.get(array[x] - x) + 1);
+			}
+			else {  // not in the map yet, put 1
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:76]: same

File: solutions/cf-1398C.mdx [hunk 11]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
+				sumDist.put(array[x] - x, (long) 1);
+			}
+		}
+		long goodArrays = 0;
+		for (int key : sumDist.keySet()) {  // for each key
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:81]: ^same

File: solutions/cf-1398C.mdx [hunk 12]
```diff
@@ -21,33 +21,85 @@ using pi = pair<ll, ll>;
...
+			long f = sumDist.get(key);
+			// this uses the formula n * (n - 1) / 2, and it gets
```
Comment (bqi343) [inline solutions/cf-1398C.mdx:83]: why not just "sum of all positive integers less than `f`"? same for cpp solution

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by nathangong — APPROVED** [review]

**Review by bqi343 — APPROVED** [review]

File: solutions/cf-1398C.mdx [hunk 13]
```diff
@@ -21,33 +21,82 @@ using pi = pair<ll, ll>;
 void solve(){
 	int n; cin >> n;
 	vector<int> dp(n+1, 0);
```
Comment (ryanchou-dev) [inline solutions/cf-1398C.mdx:23]:
```suggestion
	vector<int> dp(n + 1, 0);
```

File: solutions/cf-1398C.mdx [hunk 14]
```diff
@@ -21,33 +21,82 @@ using pi = pair<ll, ll>;
...
-	for(int i=1; i<=n; i++){
+	for (int i=1; i<=n; i++) {
```
Comment (ryanchou-dev) [inline solutions/cf-1398C.mdx:24]:
```suggestion
	for (int i = 1; i <= n; i++) {
```

File: solutions/cf-1398C.mdx [hunk 15]
```diff
@@ -21,33 +21,82 @@ using pi = pair<ll, ll>;
...
 		char c; cin >> c;
 		dp[i]=c-'0';
 	}
-	for(int i=1; i<=n; i++){
+	for (int i=1; i<=n; i++) {
```
Comment (ryanchou-dev) [inline solutions/cf-1398C.mdx:28]:
```suggestion
	for (int i = 1; i <= n; i++) {
```

File: solutions/cf-1398C.mdx [hunk 16]
```diff
@@ -21,33 +21,82 @@ using pi = pair<ll, ll>;
...
 		dp[i]+=dp[i-1];
 	}
 	map<int, ll> sumDist;
-	for(int i=0; i<=n; i++){
+	for (int i=0; i<=n; i++) {
```
Comment (ryanchou-dev) [inline solutions/cf-1398C.mdx:32]:
```suggestion
	for (int i = 0; i <= n; i++) {
```

File: solutions/cf-1398C.mdx [hunk 17]
```diff
@@ -21,33 +21,82 @@ using pi = pair<ll, ll>;
...
 	int n; cin >> n;
 	vector<int> dp(n+1, 0);
-	for(int i=1; i<=n; i++){
+	for (int i=1; i<=n; i++) {
 		char c; cin >> c;
 		dp[i]=c-'0';
 	}
-	for(int i=1; i<=n; i++){
+	for (int i=1; i<=n; i++) {
 		dp[i]+=dp[i-1];
```
Comment (ryanchou-dev) [inline solutions/cf-1398C.mdx:29]:
```suggestion
		dp[i]+=dp[i - 1];
```

File: solutions/cf-1398C.mdx [hunk 18]
```diff
@@ -21,33 +21,82 @@ using pi = pair<ll, ll>;
...
-	for(int i=0; i<=n; i++){
+	for (int i=0; i<=n; i++) {
 		sumDist[dp[i]-i]++;
 	}
 	ll goodArrays=0;
-	for(pi p: sumDist){
+	for (pi p: sumDist) {
 		ll f=p.second;
+		// sum of all positive integers less than f
 		goodArrays+=f*(f-1)/2;
```
Comment (ryanchou-dev) [inline solutions/cf-1398C.mdx:39]:
```suggestion
		goodArrays += f * (f - 1) / 2;
```

File: solutions/cf-1398C.mdx [hunk 19]
```diff
@@ -21,33 +21,82 @@ using pi = pair<ll, ll>;
...
 	}
 	cout << goodArrays << endl;
 }
 
-int main(){
+int main() {
 	int t; cin >> t;
-	for(int i=0; i<t; i++){
+	for (int i=0; i<t; i++) {
```
Comment (ryanchou-dev) [inline solutions/cf-1398C.mdx:46]:
```suggestion
	for (int i = 0; i < t; i++) {
```

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]
whitespace between ops. rest is good

**Comment by ryanchou-dev** [comment]
oh it got merged

**Comment by ryanchou-dev** [comment]
should i remove the review?

**Pull request merged**